### PR TITLE
start of Azerty layout scancode set2

### PR DESCRIPTION
**Pull request by le0kar0ub1**
Basic Azerty layout handling.
The method used for other layouts occult the azerty meaning.

### CODE REVIEW

**Comment by vinc** [comment]
Hi, I have a French azerty keyboard on my laptop so I tested your branch on my little OS (https://github.com/vinc/moros), and the letters and numbers work fine but I had trouble with some other keys.

For example the `<` key at the bottom left of the keyboard doesn't work, while the `*` key on the right behave as if it was the `<` key. Are they working correctly for you? The issue could very well comes from my OS though.

**Comment by le0kar0ub1** [comment]
Hi,

I will check this tomorrow and give you a feedback and then commit if the
error come from me.

**Comment by le0kar0ub1** [comment]
hi again,

Are we okay on the basic layout implementation ? us104 - >
https://upload.wikimedia.org/wikipedia/commons/3/3a/Qwerty.svg

Ideally i think that to be coherent on the translation to azerty layout, a
new KeyCode implementation can be a more efficient way.
Btw, as a first step i will re-map correctly when i will be certain that
the basic keyboard layout is the one i showed above. If im mistaken, can
you transfer me an image to the well layout please ?

Le sam. 4 juil. 2020 à 00:17, Vincent Ollivier <notifications@github.com> a
écrit :

**Comment by vinc** [comment]
The US Qwerty keyboard layout is 104 keys, and the UK one is 105 keys. I think the common French Azerty keyboard is 105 keys like the UK one, so it might help to reuse `Uk105Key` and code only what is changing between those two.

https://upload.wikimedia.org/wikipedia/commons/d/da/KB_United_States.svg
https://upload.wikimedia.org/wikipedia/commons/d/da/KB_United_Kingdom.svg

That's what I did when I implemented the Dvorak layout from `Us104Key` as you can see here: https://github.com/rust-embedded-community/pc-keyboard/pull/6/files#diff-ec45777dbfe740feab6d8d88592c97d1R297

**Comment by vinc** [comment]
You made me realize that actually I should also have started from `Uk105Key` to have the bottom left `<` key working on a Dvorak 105 keys keyboard!

**Comment by le0kar0ub1** [comment]
hi, I think the keymap must be okay except for the  ['<'  '>'].
I don't understand how the UK keyboard key ['|' '\'] can work, have you
tested it ?
These 2 keys trigger the same KeyCode on azerty/UK then, if it's work on
the UK this will work on the azerty.
Btw, the rest of the map should work fine.

Le sam. 4 juil. 2020 à 17:55, Vincent Ollivier <notifications@github.com> a
écrit :

**Comment by le0kar0ub1** [comment]
I havn't pull request, the code is updated on my forked version.

File: src/layouts/azerty.rs [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+            KeyCode::Fullstop => {
+                if modifiers.is_shifted() {
+                    DecodedKey::Unicode('/')
+                } else {
+                    DecodedKey::Unicode(':')
+                }
+            }
+            KeyCode::Slash => {
+                if modifiers.is_shifted() {
+                    DecodedKey::Unicode('!')
```
Comment (vinc) [inline src/layouts/azerty.rs:421]:
```suggestion
                    DecodedKey::Unicode('§')
```

File: src/layouts/azerty.rs [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+                    DecodedKey::Unicode('°')
+                } else if modifiers.alt_gr {
+                    DecodedKey::Unicode(']')
+                } else {
+                    DecodedKey::Unicode(')')
+                }
+            }
+            KeyCode::Equals => {
+                if modifiers.is_shifted() {
+                    DecodedKey::Unicode('=')
```
Comment (vinc) [inline src/layouts/azerty.rs:121]:
```suggestion
                    DecodedKey::Unicode('+')
```

File: src/layouts/azerty.rs [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+                } else if modifiers.alt_gr {
+                    DecodedKey::Unicode('}')
+                } else {
+                    DecodedKey::Unicode('+')
```
Comment (vinc) [inline src/layouts/azerty.rs:125]:
```suggestion
                    DecodedKey::Unicode('=')
```

File: src/layouts/azerty.rs [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+                    DecodedKey::Unicode('/')
+                } else {
+                    DecodedKey::Unicode(':')
+                }
+            }
+            KeyCode::Slash => {
+                if modifiers.is_shifted() {
+                    DecodedKey::Unicode('!')
+                } else {
+                    DecodedKey::Unicode('§')
```
Comment (vinc) [inline src/layouts/azerty.rs:423]:
```suggestion
                    DecodedKey::Unicode('!')
```

**Comment by le0kar0ub1** [comment]
let's go :)

**Comment by vinc** [comment]
@thejpster could you have a look at this PR?

It would be great to publish a new version of the crate after that with the new layouts :tada:

**Comment by thejpster** [comment]
I no longer have write access to this repo I'm afraid.

**Comment by thejpster** [comment]
I see I am the only person who can push to crates.io though - who would like to be added to that, and can I add a Github group to a crate?

**Comment by thejpster** [comment]
OK, once https://github.com/rust-embedded-community/meta/issues/2 is sorted, I can set the owner of pc-keyboard to be that team, and anyone in that team can push you a new release.

**Comment by vinc** [comment]
That seems great :+1:

**Pull request merged**